OpenRA.Mods.Common/Traits/AmmoPool.cs

```diff
@@ -42,99 +45,123 @@ public class AmmoPoolInfo : ITraitInfo
-		public bool HasAmmo() { return CurrentAmmo > 0; }
+		public int GetAmmoCount() { return currentAmmo; }
+		public bool FullAmmo() { return currentAmmo == Info.Ammo; }
+		public bool HasAmmo() { return currentAmmo > 0; }
 
-		public bool GiveAmmo()
+		public bool GiveAmmo(Actor self, int count)
 		{
-			if (CurrentAmmo >= Info.Ammo)
+			if (currentAmmo >= Info.Ammo || count < 0)
 				return false;
 
-			++CurrentAmmo;
+			currentAmmo = (currentAmmo + count).Clamp(0, Info.Ammo);
+
+			for (var i = 0; i < count; i++)
+				UpdateCondition(self);
+
 			return true;
 		}
 
-		public bool TakeAmmo()
+		public bool TakeAmmo(Actor self, int count)
 		{
-			if (CurrentAmmo <= 0)
+			if (currentAmmo <= 0 || count < 0)
 				return false;
 
-			--CurrentAmmo;
+			currentAmmo = (currentAmmo - count).Clamp(0, Info.Ammo);
+
+			for (var i = 0; i < count; i++)
+				UpdateCondition(self);
+
 			return true;
 		}
 
+		// This mostly serves to avoid complicated ReloadAmmoPool look-ups in various other places.
+		// TODO: Investigate removing this when the Rearm activity is replaced with a condition-based solution.
+		public bool AutoReloads { get; private set; }
+
+		void INotifyCreated.Created(Actor self)
+		{
+			conditionManager = self.TraitOrDefault<ConditionManager>();
+			AutoReloads = self.TraitsImplementing<ReloadAmmoPool>().Any(r => r.Info.AmmoPool == Info.Name && r.Info.RequiresCondition == null);
+
+			// If (Initial)Ammo is higher than 1, we need to grant an instance for each of them at creation
+			for (var i = 0; i < currentAmmo; i++)
+				UpdateCondition(self);
+
+			// HACK: Temporarily needed until Rearm activity is gone for good
+			RemainingTicks = Info.ReloadDelay;
+		}
+
 		void INotifyAttack.Attacking(Actor self, Target target, Armament a, Barrel barrel)
 		{
-			if (a != null && a.Info.AmmoPoolName == Info.Name)
-				TakeAmmo();
+			if (a != null && Info.Armaments.Contains(a.Info.Name))
+				TakeAmmo(self, 1);
 		}
 
 		void INotifyAttack.PreparingAttack(Actor self, Target target, Armament a, Barrel barrel) { }
 
-		void ITick.Tick(Actor self)
+		void GrantInstance(Actor self, string cond)
+		{
+			if (string.IsNullOrEmpty(cond))
+				return;
+
+			tokens.Push(conditionManager.GrantCondition(self, cond));
+		}
+
+		void RevokeInstance(Actor self)
 		{
-			if (!Info.SelfReloads)
+			if (tokens.Count == 0)
 				return;
 
-			// Resets the tick counter if ammo was fired.
-			if (Info.ResetOnFire && GetAmmoCount() < PreviousAmmo)
-			{
-				RemainingTicks = Info.SelfReloadDelay;
-				PreviousAmmo = GetAmmoCount();
-			}
+			conditionManager.RevokeCondition(self, tokens.Pop());
+		}
 
-			if (!FullAmmo() && --RemainingTicks == 0)
-			{
-				RemainingTicks = Info.SelfReloadDelay;
+		void UpdateCondition(Actor self)
+		{
+			if (conditionManager == null || string.IsNullOrEmpty(Info.AmmoCondition))
+				return;
 
-				for (var i = 0; i < Info.ReloadCount; i++)
-					GiveAmmo();
+			if (currentAmmo > tokens.Count && tokens.Count < Info.Ammo)
```

Comment:
Can you please make these `while` and then remove the loops that call `UpdateCondition`?